File: frontend/src/services/engagement.ts
// ═══════════════════════════════════════════
// TRIPZYY — Assist & Reviews Service
// ═══════════════════════════════════════════

import { apiClient } from "@/lib/api";
import type {
  AssistThread,
  AssistThreadStatus,
  PaginatedResponse,
  Review,
  ReviewPage,
  ReviewSubject,
  ReviewableItem,
} from "@/types";

function page(params?: { page?: number; limit?: number }) {
  return new URLSearchParams({
    page: String(params?.page ?? 1),
    limit: String(params?.limit ?? 20),
  });
}

/** Traveller-side support conversations. */
export const assistService = {
  /**
   * Start a conversation about one trip.
   *
   * `askConcierge` defaults to true so somebody mid-tour gets an answer from
   * their own trip data immediately rather than silence until office hours.
   * The reply is always labelled as the concierge and it cannot change
   * anything — the thread still reaches a human either way.
   */
  open: (
    tripId: string,
    payload: { subject: string; body: string; askConcierge?: boolean }
  ) =>
    apiClient.post<AssistThread>(`/trips/${tripId}/assist`, {
      subject: payload.subject,
      body: payload.body,
      ask_concierge: payload.askConcierge ?? true,
    }),

  list: (params?: { page?: number; limit?: number; tripId?: string }) => {
    const q = page(params);
    if (params?.tripId) q.set("trip_id", params.tripId);
    return apiClient.get<PaginatedResponse<AssistThread>>(`/assist?${q}`);
  },

  get: (threadId: string) =>
    apiClient.get<AssistThread>(`/assist/${threadId}`),

  reply: (threadId: string, body: string, askConcierge = false) =>
    apiClient.post<AssistThread>(`/assist/${threadId}/messages`, {
      body,
      ask_concierge: askConcierge,
    }),
};

/** Operator-side. Scoped server-side to the caller's own operator. */
export const operatorAssistService = {
  threads: (params?: {
    page?: number;
    limit?: number;
    status?: AssistThreadStatus;
    mineOnly?: boolean;
  }) => {
    const q = page(params);
    if (params?.status) q.set("status", params.status);
    if (params?.mineOnly) q.set("mine_only", "true");
    return apiClient.get<PaginatedResponse<AssistThread>>(
      `/operator/assist?${q}`
    );
  },

  thread: (threadId: string) =>
    apiClient.get<AssistThread>(`/operator/assist/${threadId}`),

  /** Replying claims the thread; `resolve` closes it in the same step. */
  reply: (threadId: string, body: string, resolve = false) =>
    apiClient.post<AssistThread>(`/operator/assist/${threadId}/messages`, {
      body,
      resolve,
    }),

  setStatus: (threadId: string, status: AssistThreadStatus) =>
    apiClient.put<AssistThread>(`/operator/assist/${threadId}/status`, {
      status,
    }),

  /** Pass `null` to hand it back to the unassigned pool. */
  assign: (threadId: string, memberId: string | null) =>
    apiClient.put<AssistThread>(`/operator/assist/${threadId}/assignee`, {
      member_id: memberId,
    }),
};

/**
 * Reviews.
 *
 * Worth knowing: a rating written here is not decoration — it is written back
 * onto the row the component ranker reads, so it changes what the next
 * traveller is recommended. That is why the API refuses a review of anything
 * the author did not actually book.
 */
export const reviewService = {
  create: (payload: {
    subject: ReviewSubject;
    target_id: string;
    rating: number;
    title?: string;
    body?: string;
  }) => apiClient.post<Review>("/reviews", payload),

  /** Components from tours they took, not yet rated. Drives the prompt. */
  pending: () => apiClient.get<ReviewableItem[]>("/reviews/pending"),

  mine: (params?: { page?: number; limit?: number }) =>
    apiClient.get<PaginatedResponse<Review>>(`/reviews/mine?${page(params)}`),

  /** Public: no auth needed, and the summary carries the distribution. */
  listFor: (
    subject: ReviewSubject,
    targetId: string,
    params?: { page?: number; limit?: number }
  ) =>
    apiClient.get<ReviewPage>(
      `/reviews/${subject}/${targetId}?${page(params)}`
    ),

  update: (
    reviewId: string,
    payload: { rating?: number; title?: string; body?: string }
  ) => apiClient.put<Review>(`/reviews/${reviewId}`, payload),

  remove: (reviewId: string) => apiClient.delete<null>(`/reviews/${reviewId}`),
};

File: backend/app/routers/engagement.py
"""Assist conversations and reviews, from the traveller's side.

Opening a thread hangs off a trip, because every useful answer depends on trip
context. Reviews address their subject directly — a vendor is rated as a
vendor, not as a footnote to whichever trip happened to include it.

The operator's half of assist lives in ``routers/operator.py``, scoped to the
caller's own operator.
"""

import uuid
from typing import Annotated

from fastapi import APIRouter, Path, Query

from app.core import responses
from app.core.deps import CurrentUser, DbSession, Pagination
from app.models.enums import ReviewSubject
from app.schemas.engagement import (
    AssistThreadResponse,
    MessageRequest,
    ReviewCreateRequest,
    ReviewResponse,
    ReviewUpdateRequest,
    ReviewableItem,
    ThreadOpenRequest,
)
from app.services.assist_service import AssistService
from app.services.review_service import ReviewService

trip_assist_router = APIRouter(prefix="/trips", tags=["assist"])
assist_router = APIRouter(prefix="/assist", tags=["assist"])
reviews_router = APIRouter(prefix="/reviews", tags=["reviews"])


# --------------------------------------------------------------------------
# Assist
# --------------------------------------------------------------------------

@trip_assist_router.post(
    "/{trip_id}/assist", summary="Ask for help on a trip", status_code=201
)
async def open_thread(
    trip_id: Annotated[uuid.UUID, Path()],
    payload: ThreadOpenRequest,
    current_user: CurrentUser,
    db: DbSession,
):
    """Start a conversation with the people running your tour.

    It is routed to the operator behind the booking and their coordinators. By
    default the AI concierge also answers immediately from your trip's own
    data — labelled as the concierge, never as a person, and it cannot change
    anything. The thread stays open for a human either way.
    """
    thread = await AssistService(db).open_thread(
        trip_id,
        current_user,
        subject=payload.subject,
        body=payload.body,
        ask_concierge=payload.ask_concierge,
    )
    return responses.success(
        AssistThreadResponse(**AssistService.serialise(thread)).model_dump(),
        "Conversation started",
        status_code=201,
    )


@assist_router.get("", summary="Your conversations")
async def list_threads(
    current_user: CurrentUser,
    db: DbSession,
    pagination: Pagination,
    trip_id: Annotated[uuid.UUID | None, Query()] = None,
):
    items, total = await AssistService(db).list_for_traveller(
        current_user,
        offset=pagination.offset,
        limit=pagination.limit,
        trip_id=trip_id,
    )
    return responses.paginated(
        [
            AssistThreadResponse(
                **AssistService.serialise(t, include_messages=False)
            ).model_dump()
            for t in items
        ],
        page=pagination.page,
        limit=pagination.limit,
        total=total,
    )


@assist_router.get("/{thread_id}", summary="One conversation in full")
async def get_thread(
    thread_id: Annotated[uuid.UUID, Path()],
    current_user: CurrentUser,
    db: DbSession,
):
    thread = await AssistService(db).get_for_traveller(thread_id, current_user)
    return responses.success(
        AssistThreadResponse(**AssistService.serialise(thread)).model_dump(), "OK"
    )


@assist_router.post("/{thread_id}/messages", summary="Reply", status_code=201)
async def reply(
    thread_id: Annotated[uuid.UUID, Path()],
    payload: MessageRequest,
    current_user: CurrentUser,
    db: DbSession,
):
    thread = await AssistService(db).reply_as_traveller(
        thread_id,
        current_user,
        payload.body,
        ask_concierge=payload.ask_concierge,
    )
    return responses.success(
        AssistThreadResponse(**AssistService.serialise(thread)).model_dump(),
        "Message sent",
        status_code=201,
    )


# --------------------------------------------------------------------------
# Reviews
# --------------------------------------------------------------------------

@reviews_router.post("", summary="Review something you travelled on", status_code=201)
async def create_review(
    payload: ReviewCreateRequest,
    current_user: CurrentUser,
    db: DbSession,
):
    """Rate a service, vendor, operator or your own trip.

    Refused unless you hold a booking that actually contains it. This rating is
    written back onto the row the component ranker reads, so an unverifiable
    one would not merely be noise — it would change what the next traveller is
    recommended.
    """
    review = await ReviewService(db).create(
        current_user,
        subject=payload.subject,
        target_id=payload.target_id,
        rating=payload.rating,
        title=payload.title,
        body=payload.body,
    )
    return responses.success(
        ReviewResponse(**ReviewService.serialise(review)).model_dump(),
        "Thanks — your review is live",
        status_code=201,
    )


@reviews_router.get("/pending", summary="What you could still review")
async def pending_reviews(current_user: CurrentUser, db: DbSession):
    """Components from tours you actually took, that you have not yet rated."""
    rows = await ReviewService(db).reviewable(current_user)
    return responses.success(
        [ReviewableItem(**r).model_dump() for r in rows], "OK"
    )


@reviews_router.get("/mine", summary="Reviews you have written")
async def my_reviews(
    current_user: CurrentUser, db: DbSession, pagination: Pagination
):
    items, total = await ReviewService(db).list_by_author(
        current_user, offset=pagination.offset, limit=pagination.limit
    )
    return responses.paginated(
        [ReviewResponse(**ReviewService.serialise(r)).model_dump() for r in items],
        page=pagination.page,
        limit=pagination.limit,
        total=total,
    )


@reviews_router.get(
    "/{subject}/{target_id}", summary="Reviews of one thing, with its rating"
)
async def list_reviews(
    subject: Annotated[ReviewSubject, Path()],
    target_id: Annotated[uuid.UUID, Path()],
    db: DbSession,
    pagination: Pagination,
):
    """Public. The summary carries the distribution as well as the average."""
    items, total, summary = await ReviewService(db).list_for(
        subject, target_id, offset=pagination.offset, limit=pagination.limit
    )
    payload = {
        "items": [
            ReviewResponse(**ReviewService.serialise(r)).model_dump() for r in items
        ],
        "pagination": responses.PaginationMeta.build(
            page=pagination.page, limit=pagination.limit, total=total
        ).model_dump(),
        "summary": summary,
    }
    return responses.success(payload, "OK")


@reviews_router.put("/{review_id}", summary="Edit your review")
async def update_review(
    review_id: Annotated[uuid.UUID, Path()],
    payload: ReviewUpdateRequest,
    current_user: CurrentUser,
    db: DbSession,
):
    review = await ReviewService(db).update(
        review_id,
        current_user,
        rating=payload.rating,
        title=payload.title,
        body=payload.body,
    )
    return responses.success(
        ReviewResponse(**ReviewService.serialise(review)).model_dump(),
        "Review updated",
    )


@reviews_router.delete("/{review_id}", summary="Delete your review")
async def delete_review(
    review_id: Annotated[uuid.UUID, Path()],
    current_user: CurrentUser,
    db: DbSession,
):
    """Removing it also recomputes the aggregate it was contributing to."""
    await ReviewService(db).delete(review_id, current_user)
    return responses.success(None, "Review deleted")
